didcomm/message.py

```diff
@@ -0,0 +1,96 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Optional, List, Union, Dict
+
+from didcomm.common.types import JSON_DATA, DID, DID_URL
+
+Header = Dict[str, Union[str, int, JSON_DATA]]
+
+
+@dataclass
+class MessageOptionalHeaders:
+    """Optional headers for a message"""
+    frm: Optional[DID] = None
+    to: Optional[List[DID]] = None
+    created_time: Optional[int] = None
+    expires_time: Optional[int] = None
+    from_prior: Optional[FromPrior] = None
+    please_ack: Optional[bool] = None
+    ack: Optional[List[str]] = None
+    thid: Optional[str] = None
+    pthid: Optional[str] = None
+    attachments: Optional[List[Attachment]] = None
+    custom_headers: Optional[List[Header]] = None
+
+
+@dataclass
+class MessageRequiredHeaders:
+    """Required headers for a message"""
+    id: str
+    type: str
+    typ: str = "application/didcomm-plain+json"
+
+
+@dataclass
+class MessageBody:
+    """Message body as a application/protocol specific data"""
+    body: JSON_DATA
+
+
+@dataclass
+class Message(MessageOptionalHeaders, MessageRequiredHeaders, MessageBody):
+    """
+    Message consisting of headers and application/protocol specific data (body).
+    In order to convert a message to a DIDComm message for further transporting, call one of the following:
+    - `pack_encrypted` to build an Encrypted DIDComm message
+    - `pack_signed` to build a signed DIDComm message
+    - `pack_plaintext` to build a Plaintext DIDComm message
+    """
+    pass
+
+
+@dataclass(frozen=True)
+class Attachment:
+    """Plaintext attachment"""
+    id: str
+    data: Union[AttachmentDataLinks, AttachmentDataBase64, AttachmentDataJson]
+    description: Optional[str] = None
+    filename: Optional[str] = None
```

Comment:
'filename' MUST be present 
https://identity.foundation/didcomm-messaging/spec/#reference-2

Reply:
https://identity.foundation/didcomm-messaging/spec/#reference-2
> filename: A hint about the name that might be used if this attachment is persisted as a file. **It is not required,** and need not be unique.